File: OkHttpRequset/library/src/main/java/com/lee/okhttp/interceptor/RetryInterceptor.java
package com.lee.okhttp.interceptor;

import android.util.Log;

import com.lee.okhttp.OkHttpClient;
import com.lee.okhttp.core.InterceptorManager;
import com.lee.okhttp.core.Response;

import java.io.IOException;

/**
 * @author jv.lee
 * @date 2019-08-29
 * @description 重试拦截器
 */
public class RetryInterceptor implements Interceptor {

    private static final String TAG = "OkHttp";

    private OkHttpClient client;

    public RetryInterceptor(OkHttpClient client) {
        this.client = client;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        InterceptorManager chainManager = (InterceptorManager) chain;
        IOException ioException = null;

        if (client.getRetryCount() != 0) {
            for (int i = 0; i < client.getRetryCount(); i++) {
                try {
                    Log.i(TAG, "intercept: 重试拦截器 返回response");

                    //如果没有异常循环就结束了
                    return chain.proceed(chainManager.request());
                } catch (IOException e) {
                    ioException = e;
                }
            }
        }
        assert ioException != null;
        throw ioException;
    }
}

File: KotlinLibraryX/library/src/main/java/com/lee/library/net/interceptor/RetryInterceptor.kt
package com.lee.library.net.interceptor

import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

/**
 * 重试拦截器
 * @author jv.lee
 * @date 2021/11/1
 */
class RetryInterceptor(
    var maxRetry: Int
) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        var response = chain.proceed(request)
        var retryNum = 0
        while (!response.isSuccessful && retryNum < maxRetry) {
            retryNum++
            response = chain.proceed(request)
        }
        return response
    }
}